hash-parser/src/hash.pest

```diff
@@ -2,443 +2,453 @@
+
+/// Typed expression (as)
+typed_expr = { unary_expr ~ ( as_kw ~ any_type )? }
+
+/// Binary expression
+binary_expr = { typed_expr ~ ( binary_op ~ typed_expr )* }
+
+/// Block expression
+block_expr = { block }
+
+/// Any expression
+expr = { block_expr | binary_expr }
+
+/// ---- Patterns ----
+
+/// Binding pattern
+binding_pattern = { name }
+
+/// Struct pattern
+struct_pattern_field = { name ~ (eq_op ~ pattern)? }
+struct_pattern_fields = { curly_left ~ (struct_pattern_field ~ (comma_op ~ struct_pattern_field)*)? ~ curly_right }
+struct_pattern = { access_name ~ struct_pattern_fields }
+
+/// Enum pattern
+enum_pattern_args = { paren_left ~ pattern ~ (comma_op ~ pattern)* ~ paren_right }
+enum_pattern = { access_name ~ enum_pattern_args }
+
+/// Namespace pattern
+namespace_pattern_field = { name ~ (eq_op ~ pattern)? }
+namespace_pattern = { curly_left ~ (namespace_pattern_field ~ (comma_op ~ namespace_pattern_field)*)? ~ curly_right }
+
+/// Tuple pattern
+tuple_pattern = { paren_left ~ (comma_op ~ &paren_right | pattern ~ comma_op ~ &paren_right | (pattern ~ (comma_op ~ pattern)* ~ comma_op?)) ~ paren_right }
 
+/// Ignore pattern
+ignore_pattern = { underscore_op }
 
-braced_pattern = _{ "(" ~ pattern ~ ")" }
-ignore_pattern = _{"_"}
+/// Grouped pattern (1 | 2)
+grouped_pattern = { paren_left ~ pattern ~ paren_right }
 
-/// Single pattern combinator is used as a combinator for match
-/// statements where multiple patterns can be specified for a 
-/// switch case. 
-single_pattern = { 
-    braced_pattern     |
-    enum_pattern       |
-    exhaustive_pattern |
-    ignore_pattern     |
-    literal
+/// Single pattern (used in binding positions)
+single_pattern = {
+    ignore_pattern  |
+    binding_pattern |
+    enum_pattern    |
+    struct_pattern  |
+    tuple_pattern   |
+    grouped_pattern |
+    namespace_pattern 
 }
 
-/// The pattern grammar
-pattern = {single_pattern ~ ("if" ~ expr)? ~ ("|" ~ pattern)?}
+/// Compound pattern (matches a series of single patterns)
+compound_pattern = { single_pattern ~ (pipe_op ~ single_pattern)* }
 
-/// Conditional if expression grammar
-if_statement = {"if" ~ expr ~ block ~ ("else" ~ (if_statement | block))?}
+/// Any pattern (includes optional if condition)
+pattern = { compound_pattern ~ (if_kw ~ expr)? }
 
-/// Conditional match expression grammar
-match_case      = {pattern ~ "=>" ~ expr}
-match_statement = {"match" ~ expr ~ "{" ~ (match_case ~ semi)* ~ "}"}
+/// ---- Blocks ----
 
-/// Body block grammar, body blocks can be used within any kind
-/// of expression to denote a body of a particular expression.
-/// The expression at the end is used as a implicit 'return' statement
-/// from the given body block.
-body_block = { "{" ~  statement* ~ expr? ~ "}" }
+/// Body block
+/// Zero or one statements followed by zero or one expression.
+body_block = { curly_left ~ statement* ~ expr? ~ curly_right } // @Broken: statement parsing needs to be non-greedy
```

Comment:
Yes I was hoping we could do the transformation at AST level since pest is by nature greedy

Reply:
Yeah, there's probably no other way to do it from within pest.